console/collections/src/kary_merkle_tree/mod.rs

```diff
@@ -89,28 +89,55 @@ impl<LH: LeafHash<Hash = PH::Hash>, PH: PathHash, const DEPTH: u8, const ARITY:
         // Compute the empty hash.
         let empty_hash = path_hasher.hash_empty::<ARITY>()?;
 
+        // Calculate the size of the tree which excludes leafless nodes.
+        let arity = ARITY as usize;
+        let minimum_tree_size = std::cmp::max(
+            1,
+            num_nodes + leaves.len() + if leaves.len() % arity == 0 { 0 } else { arity - leaves.len() % arity },
+        );
+
         // Initialize the Merkle tree.
-        let mut tree = vec![empty_hash; tree_size];
+        let mut tree = vec![empty_hash; minimum_tree_size];
 
         // Compute and store each leaf hash.
         tree[num_nodes..num_nodes + leaves.len()].clone_from_slice(&leaf_hasher.hash_leaves(leaves)?);
         lap!(timer, "Hashed {} leaves", leaves.len());
 
         // Compute and store the hashes for each level, iterating from the penultimate level to the root level.
         let mut start_index = num_nodes;
+        let mut current_empty_node_hash = path_hasher.hash_children(&vec![empty_hash; arity])?;
         // Compute the start index of the current level.
         while let Some(start) = parent::<ARITY>(start_index) {
             // Compute the end index of the current level.
             let end = child_indexes::<ARITY>(start).next().ok_or_else(|| anyhow!("Missing left-most child"))?;
 
             // Construct the children for each node in the current level.
             let child_nodes = (start..end)
-                .map(|i| child_indexes::<ARITY>(i).map(|child_index| tree[child_index]).collect::<Vec<_>>())
+                .map(|i| {
+                    // Prepare an iterator over then children, being mindful of possible missing leaves.
+                    let child_iter = || child_indexes::<ARITY>(i).map(|child_index| tree.get(child_index).copied());
+
+                    // At the leaf level just return the leaves, or `None` in case the node has none.
+                    if current_empty_node_hash == empty_hash {
+                        return child_iter().collect::<Option<Vec<_>>>();
+                    }
+
+                    // Check if the children aren't all derived from empty hashes.
+                    if child_iter().all(|hash| hash == Some(current_empty_node_hash)) {
```

Comment:
Does the optimized pathway ever get triggered for any level past the first?

Reply:
It doesn't; just like with your previous comment, I think this was done before I realized that the empty hashes don't actually need to be there.

Reply:
Can't see what it does anymore, but removing it is a win for simplicity anyway